src/liquidations.ts

```diff
@@ -40,20 +40,129 @@ export function isBadDebt(user: PositionsEstimate): boolean {
  * @param user - The positions estimate of the user
  * @returns The liquidation percent
  */
-export function calculateLiquidationPercent(user: PositionsEstimate): number {
-  const avgInverseLF = user.totalEffectiveLiabilities / user.totalBorrowed;
-  const avgCF = user.totalEffectiveCollateral / user.totalSupplied;
-  const estIncentive = 1 + (1 - avgCF / avgInverseLF) / 2;
-  const numerator = user.totalEffectiveLiabilities * 1.06 - user.totalEffectiveCollateral;
-  const denominator = avgInverseLF * 1.06 - avgCF * estIncentive;
-  const liqPercent = Math.min(
-    Math.round((numerator / denominator / user.totalBorrowed) * 100),
-    100
+export function calculateLiquidation(
+  pool: Pool,
+  user: Positions,
+  estimate: PositionsEstimate,
+  oracle: PoolOracle
+): {
+  auctionPercent: number;
+  lot: string[];
+  bid: string[];
+} {
+  let effectiveCollaterals: [number, number][] = [];
+  let rawCollaterals: Map<string, number> = new Map();
+  let effectiveLiabilities: [number, number][] = [];
+  let rawLiabilities: Map<string, number> = new Map();
+
+  for (let [index, amount] of user.collateral) {
+    let assetId = pool.metadata.reserveList[index];
+    let oraclePrice = oracle.getPriceFloat(assetId);
+    let reserve = pool.reserves.get(assetId);
+    if (oraclePrice === undefined || reserve === undefined) {
+      continue;
+    }
+    let effectiveAmount = reserve.toEffectiveAssetFromBTokenFloat(amount) * oraclePrice;
+    let rawAmount = reserve.toAssetFromBTokenFloat(amount) * oraclePrice;
+    effectiveCollaterals.push([index, effectiveAmount]);
+    rawCollaterals.set(assetId, rawAmount);
+  }
+  for (let [index, amount] of user.liabilities) {
+    let assetId = pool.metadata.reserveList[index];
+    let oraclePrice = oracle.getPriceFloat(assetId);
+    let reserve = pool.reserves.get(assetId);
+    if (oraclePrice === undefined || reserve === undefined) {
+      continue;
+    }
+    let effectiveAmount = reserve.toEffectiveAssetFromDTokenFloat(amount) * oraclePrice;
+    let rawAmount = reserve.toAssetFromDTokenFloat(amount) * oraclePrice;
+    effectiveLiabilities.push([index, effectiveAmount]);
+    rawLiabilities.set(assetId, rawAmount);
+  }
+
+  effectiveCollaterals.sort((a, b) => b[1] - a[1]);
+  effectiveLiabilities.sort((a, b) => b[1] - a[1]);
+  let firstCollateral = effectiveCollaterals.shift();
+  let firstLiability = effectiveLiabilities.shift();
```

Comment:
(picky) shift causes a bunch of memory allocations. It doesn't really matter here, but we might be better off doing it the more "normal" way.

sort the array least to greatest, and pop elements off the end